deer/fsolve_ivp.py

```diff
@@ -165,3 +166,230 @@ def solve_ivp_inv_lin(self, gmat: List[jnp.ndarray], rhs: jnp.ndarray,
+
+
+class RK3(GeneralODE):
+    """
+    Compute the solution of initial value problem with the Runge-Kutta 3rd order method.
+
+    Arguments
+    ---------
+    step_size: float
+        The step size to use for the RK3 method. If None, it will use (tpts[i] - tpts[i - 1]).
+    """
+
+    def ode_step(
+        self,
+        func: Callable[[jnp.ndarray, jnp.ndarray, Any], jnp.ndarray], 
+        yi: jnp.ndarray, 
+        xi: jnp.ndarray, 
+        dt: jnp.ndarray, 
+        dx: jnp.ndarray, 
+        params: Any
+    ) -> Tuple[jnp.ndarray, jnp.ndarray]:
+        k1 = func(yi, xi, params)
+        k2 = func(yi + dt * k1, xi + dx, params)
+        k3 = func(yi + 0.25 * dt * (k1 + k2), xi + 0.5 * dx, params)
+
+        yi = yi + (dt / 6.0) * (k1 + k2 + 4 * k3)
+        return yi
+
+
+class RK4(GeneralODE):
+    """
+    Compute the solution of initial value problem with the Runge-Kutta 4th order method.
+
+    Arguments
+    ---------
+    step_size: float
+        The step size to use for the RK4 method. If None, it will use (tpts[i] - tpts[i - 1]).
+    """
+    
+    def ode_step(
+        self,
+        func: Callable[[jnp.ndarray, jnp.ndarray, Any], jnp.ndarray], 
+        yi: jnp.ndarray, 
+        xi: jnp.ndarray, 
+        dt: jnp.ndarray, 
+        dx: jnp.ndarray, 
+        params: Any
+    ) -> Tuple[jnp.ndarray, jnp.ndarray]:
+        k1 = func(yi, xi, params)
+        k2 = func(yi + 0.5 * dt * k1, xi + 0.5 * dx, params)
+        k3 = func(yi + 0.5 * dt * k2, xi + 0.5 * dx, params)
+        k4 = func(yi + dt * k3, xi + dx, params)
+
+        yi = yi + (dt / 6.0) * (k1 + 2*k2 + 2*k3 + k4)
+        return yi
+
+
+class GeneralBackwardODE(GeneralODE):
+    """
+    Compute the solution of initial value problem with Backward ODE methods.
+
+    Arguments
+    ---------
+    step_size: float
+        The step size for ODE solver. If None, it will use (tpts[i] - tpts[i - 1]).
+    tol: float
+        The tolerance for the fixed-point iteration.
+    max_iter: int
+        The maximum number of iterations for the fixed-point iteration.
+    """
+    def __init__(self, tol: float = 1e-6, max_iter: int = 100):
+        self.tol = tol
+        self.max_iter = max_iter
+
+
+class BackwardEuler(GeneralBackwardODE):
+    """
+    Compute the solution of initial value problem with the Backward Euler method.
+    """
+    
+    def ode_step(
+        self,
+        func: Callable[[jnp.ndarray, jnp.ndarray, Any], jnp.ndarray],
+        yi: jnp.ndarray,
+        xi: jnp.ndarray,
+        dt: jnp.ndarray,
+        dx: jnp.ndarray,
+        params: Any
+    ) -> Tuple[jnp.ndarray, jnp.ndarray]:
+        
+        def cond(val):
+            y_new, y_new_next, i = val
+            return jnp.logical_and(i < self.max_iter, jnp.linalg.norm(y_new_next - y_new) >= self.tol)
+        
+        def body(val):
+            y_new, _, i = val
+            y_new_next = yi + dt * func(y_new, xi, params)
+            return y_new_next, y_new_next, i + 1
```

Comment:
This might not converge on some functions. You can use root finder here to improve the convergence. You can see this one to get an example on how to use the rootfinder in backward euler https://github.com/machine-discovery/deer/blob/ae4a864071937a69729ab5cc447cc6bbdf53af7f/deer/fsolve_idae.py#L108-L123

Reply:
Added for backward ode methods